File: collector/filters.py
"""Title filter. Mirrors docs/filters.js exactly; tests/filter_cases.json keeps them in sync.

Matching rule: a keyword of 6 or more characters matches anywhere in the title
(so 'planner' finds 'Transportplanner'). A shorter keyword must start a word
(so 'rail' finds 'Rail Planner' but not 'Retail').
"""
import re

from .text import fold

LONG_KEYWORD = 6


def keyword_matches(folded_title, keyword):
    kw = fold(keyword).strip()
    if not kw:
        return False
    if len(kw) >= LONG_KEYWORD:
        return kw in folded_title
    return re.search(r"(^|[^a-z0-9])" + re.escape(kw), folded_title) is not None


def any_match(title, keywords):
    folded = fold(title)
    return any(keyword_matches(folded, k) for k in keywords)


def is_suitable(title, filters, extra_include=()):
    include = list(filters.get("include", [])) + list(extra_include)
    return any_match(title, include) and not any_match(title, filters.get("exclude", []))


def is_air_or_ocean_only(title, filters):
    return any_match(title, filters.get("air_ocean", [])) and not any_match(title, filters.get("land", []))

File: docs/filters.js
// Title filter. Mirrors collector/filters.py exactly; tests/filter_cases.json keeps them in sync.
// A keyword of 6+ characters matches anywhere in the title; a shorter one must start a word.

const LONG_KEYWORD = 6;

export function fold(text) {
  return (text || "").normalize("NFD").replace(/[\u0300-\u036f]/g, "").toLowerCase();
}

function escapeRegex(text) {
  return text.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

export function keywordMatches(foldedTitle, keyword) {
  const kw = fold(keyword).trim();
  if (!kw) return false;
  if (kw.length >= LONG_KEYWORD) return foldedTitle.includes(kw);
  return new RegExp("(^|[^a-z0-9])" + escapeRegex(kw)).test(foldedTitle);
}

export function anyMatch(title, keywords) {
  const folded = fold(title);
  return (keywords || []).some((k) => keywordMatches(folded, k));
}

export function isSuitable(title, filters, extraInclude = []) {
  const include = [...(filters.include || []), ...extraInclude];
  return anyMatch(title, include) && !anyMatch(title, filters.exclude || []);
}

export function isAirOrOceanOnly(title, filters) {
  return anyMatch(title, filters.air_ocean || []) && !anyMatch(title, filters.land || []);
}
